File: server-dev/models/recipe.py
from typing import List, Optional

from bson import ObjectId
from pydantic import BaseModel, Field, ConfigDict

from models.mongo import PyObjectId
from models.shared import LocalizedString


class Author(BaseModel):
    name: str
    url: Optional[str] = Field(alias="url", default=None)


class CookBook(BaseModel):
    name: Optional[str] = Field(alias="name", default=None)
    url: str
    publisher: Optional[str] = Field(alias="publisher", default=None)


class Source(BaseModel):
    authors: List[Author] = Field(alias="authors", default=[])
    url: Optional[str] = Field(alias="url", default=None)
    cookBook: Optional[CookBook] = Field(alias="cookBook", default=None)


class Step(BaseModel):
    desc: LocalizedString = Field(alias="desc", default=LocalizedString(lang="en", value=""))
    items: Optional[List[str]] = Field(alias="items", default=[])
    imgUrl: Optional[str] = Field(alias="imgUrl", default=None)
    duration: Optional[int] = Field(alias="duration", default=None, ge=1)
    temperature: Optional[float] = Field(alias="temperature", default=None)

    @staticmethod
    def new(desc: LocalizedString, items: List[str], imgUrl: str, duration: int, temperature: float):
        return Step(desc=desc, items=items, imgUrl=imgUrl, duration=duration, temperature=temperature)


class RecipeItem(BaseModel):
    id: PyObjectId = Field(alias="id", default=None)
    quantity: Optional[float] = Field(alias="quantity", default=None, ge=0)
    unit: Optional[str] = Field(alias="unit", default=None)

    @staticmethod
    def new(item_id: ObjectId, quantity: float, unit: str):
        return RecipeItem(id=item_id, quantity=quantity, unit=unit)


class Recipe(BaseModel):
    id: Optional[PyObjectId] = Field(alias="_id", default=None)
    name: LocalizedString = Field(alias="name", default=LocalizedString(lang="en", value=""))
    desc: LocalizedString = Field(alias="desc", default=LocalizedString(lang="en", value=""))
    items: List[RecipeItem] = Field(alias="items", default=[])
    steps: List[Step] = Field(alias="steps", default=[])
    props: dict = Field(alias="props", default={})
    src: Optional[Source] = Field(alias="src", default=None)
    deleted: bool = Field(alias="deleted", default=False)

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_schema_extra={
            "example": {
                "name": {
                    "lang": "en",
                    "value": "Example Recipe"
                },
                "desc": {
                    "lang": "en",
                    "value": "This is an example recipe."
                },
                "items": [
                    {
                        "id": "507f1f77bcf86cd799439011",
                        "quantity": 1,
                        "unit": "cup"
                    }
                ],
                "steps": [
                    {
                        "desc": {
                            "lang": "en",
                            "value": "This is an example step."
                        },
                        "items": [
                            "507f1f77bcf86cd799439011"
                        ],
                        "imgUrl": "https://example.com/example.jpg",
                        "duration": 10,
                        "temperature": 350
                    }
                ],
                "props": {
                    "prepTime": 10,
                    "cookTime": 30,
                    "servings": 4,
                    "tags": [
                        "example",
                        "recipe"
                    ]
                },
                "src": {
                    "url": "https://example.com/example.html",
                    "authors": [
                        {
                            "name": "John Smith",
                            "url": "https://example.com/john-smith"
                        }
                    ],
                    "cookBook": {
                        "name": "Example Cookbook",
                        "url": "https://example.com/cookbook",
                        "publisher": "Example Publisher"
                    }
                }
            }
        }
    )

File: server-prod/recipe.go
/*
Copyright © 2023 JOSEF MUELLER
*/
package main

import (
	"errors"
	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"

	"go.mongodb.org/mongo-driver/bson"
)

type Recipe struct {
	ID          primitive.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	Name        LocalizedString    `json:"name,omitempty" bson:"name,omitempty"`
	Description LocalizedString    `json:"desc" bson:"desc,omitempty"`
	Items       []RecipeItem       `json:"items" bson:"items" binding:"required"`
	Steps       []Step             `json:"steps" bson:"steps" binding:"required"`
	Props       struct {
		ImgUrl    string    `json:"imgUrl,omitempty" bson:"imgUrl,omitempty"`
		Duration  int       `json:"duration,omitempty" bson:"duration,omitempty"`
		CreatedAt time.Time `json:"date,omitempty" bson:"date,omitempty"`
		Tags      []string  `json:"tags,omitempty" bson:"tags,omitempty"`
	} `json:"props,omitempty" bson:"props,omitempty"`
	Source struct {
		Url     string `json:"url,omitempty" bson:"url,omitempty"`
		Authors []struct {
			Name string `json:"name" bson:"name" binding:"required"`
			Url  string `json:"url,omitempty" bson:"url,omitempty"`
		} `json:"authors,omitempty" bson:"authors,omitempty"`
		Copyright string `json:"cr,omitempty" bson:"cr,omitempty"`
		CookBook  struct {
			Name      string `json:"name,omitempty" bson:"name,omitempty"`
			Url       string `json:"url,omitempty" bson:"url,omitempty"`
			Publisher string `json:"pub,omitempty" bson:"pub,omitempty"`
		} `json:"cookBook,omitempty" bson:"cookBook,omitempty"`
	} `json:"src,omitempty" bson:"src,omitempty"`
	Deleted bool `json:"-" bson:"deleted,omitempty"`
}

type RecipeItem struct {
	ID       primitive.ObjectID `json:"id,omitempty" bson:"id,omitempty"`
	Quantity float64            `json:"quantity" bson:"quantity" binding:"required"`
	Unit     string             `json:"unit,omitempty" bson:"unit,omitempty" binding:"required"`
	Notes    LocalizedString    `json:"notes,omitempty" bson:"notes,omitempty"`
}

type Step struct {
	Description LocalizedString `json:"desc" bson:"desc" binding:"required"`
	ImgUrl      string          `json:"imgUrl,omitempty" bson:"imgUrl,omitempty"`
	Duration    int             `json:"duration,omitempty" bson:"duration,omitempty"`
	Temperature float32         `json:"temperature,omitempty" bson:"temperature,omitempty"`
}

// HandleGetRecipes gets called by server
// Calls getRecipesFromDB and handles the context
func (server *TasteBuddyServer) HandleGetRecipes(context *fiber.Ctx) error {
	recipes, err := server.GetRecipes(true)
	if err != nil {
		server.LogError("HandleGetRecipes.GetRecipes", err)
		return ErrorMessage(context, "Could not get recipes")
	}
	return Success(context, recipes)
}

var (
	cachedRecipes    []Recipe
	cachedRecipesAge int64
)

// GetRecipesCollection gets recipes collection from database
func (app *TasteBuddyApp) GetRecipesCollection() *TBCollection {
	return app.GetDBCollection("recipes")
}

// UpdateRecipeCache updates the recipe cache
func (app *TasteBuddyApp) UpdateRecipeCache() error {
	app.LogTrace("UpdateRecipeCache", "Update recipe cache")
	if _, err := app.GetRecipes(false); err != nil {
		return app.LogError("UpdateRecipeCache.GetRecipes", err)
	}
	return nil
}

// GetRecipes gets all recipes from database
func (app *TasteBuddyApp) GetRecipes(cached bool) ([]Recipe, error) {
	if cachedRecipes == nil || !cached || time.Now().Unix()-cachedRecipesAge > 3600 {
		app.Log("GetRecipes", "Get all recipes from database")
		var recipesFromDatabase = make([]Recipe, 0)
		// get all recipes from database that are not deleted
		if err := app.GetRecipesCollection().AllWithDefault(bson.M{"deleted": bson.M{"$ne": true}}, &recipesFromDatabase, []Recipe{}); err != nil {
			return []Recipe{}, app.LogError("GetRecipes.AllWithDefault", err)
		}
		cachedRecipes = recipesFromDatabase
		cachedRecipesAge = time.Now().Unix()
	} else {
		app.Log("GetRecipes", "Get all recipes from cache")
	}

	if len(cachedRecipes) == 0 {
		cachedRecipes = []Recipe{}
		return cachedRecipes, app.LogError("GetRecipes", errors.New("no recipes found"))
	}

	return cachedRecipes, nil
}
